tests/cpp/test_matmul_scheduler.cpp

```diff
@@ -2238,16 +2234,363 @@ TEST_F(MatmulSchedulerTest, StridedBatchEpilogueSingleBias) {
+        // Output always fully vectorizable
+        run(504,
+            136,
+            249,
+            8,
+            8,
+            8,
+            contig_K_A ? 1 : 8,
+            contig_K_B ? 1 : 8,
+            max_vec_epi);
+        // Odd N. Output not vectorizable. A fully vectorizable. B fully
+        // vectorizable unless N is the contiguous dim.
+        run(504, 137, 248, 8, 8, 8, 8, contig_K_B ? 8 : 1, 1);
+        // Odd M. Output fully vectorizable. B fully vectorizable. A fully
+        // vectorizable unless M is the contiguous dim.
+        run(505, 136, 248, 8, 8, 8, contig_K_A ? 8 : 1, 8, max_vec_epi);
+        // Odd M and N. Output not vectorizable. A and B fully vectorizable
+        // unless K is not the contiguous dim.
+        run(505, 137, 248, 8, 8, 8, contig_K_A ? 8 : 1, contig_K_B ? 8 : 1, 1);
+        // Odd M, N, K. None vectorizable.
+        run(505, 137, 249, 8, 8, 8, 1, 1, 1);
+        // Cases with vectorizable strides but misaligned base pointers
+        // A not vectorizable due to pointer offset
+        run(504, 136, 248, 2, 8, 8, 2, 8, max_vec_epi);
+        // B not vectorizable due to pointer offset
+        run(504, 136, 248, 8, 2, 8, 8, 2, max_vec_epi);
+        // epilogue not vectorizable due to pointer offset
+        // Disabled temporarily: https://github.com/NVIDIA/Fuser/issues/2169
+        // run(504, 136, 248, 8, 8, 2, 8, 8, 2);
+      }
+    }
+  }
+}
+
+// Test matmul with strided inputs. This tests that vectorization is properly
+// computed.
+TEST_F(MatmulSchedulerTest, StridedInputs) {
+  NVFUSER_TEST_CUDA_ARCH_RANGE_GUARD(7, 5, 9, 0);
+  for (auto layout : kAllSupportedMmaLayout) {
+    for (bool add_2d_bias : {false, true}) {
+      for (bool downcast_output : {false, true}) {
+        auto run = [&](int M,
+                       int N,
+                       int K,
+                       // Pointer alignment
+                       int align_A,
+                       int align_B,
+                       int align_bias,
+                       // Padding
+                       int pad_A,
+                       int pad_B,
+                       int pad_bias,
+                       int expected_vec_A,
+                       int expected_vec_B,
+                       int expected_vec_epilogue) {
+          auto fusion = std::make_unique<Fusion>();
+          FusionGuard fg(fusion.get());
+
+          // Inputs are contiguous in their inner dimension but discontiguous
+          // in the outer dim.
+          auto tv0 = TensorViewBuilder()
+                         .ndims(2)
+                         .contiguity({false, true})
+                         .dtype(DataType::Half)
+                         .build();
+          auto tv1 = TensorViewBuilder()
+                         .ndims(2)
+                         .contiguity({false, true})
+                         .dtype(DataType::Half)
+                         .build();
```

Comment:
Currently, our matmul scheduler will accept a matmul segment whose inputs are marked discontigous. Then if contiguous inputs are given, we set the vectorization factor and wind up vectorizing. This leads to a validation error in the fusion since the `TensorView` was not marked contiguous. It is rare that this would happen in practice, but we could check that the fusion's input TensorViews have contiguous inner dimensions when inferring vectorization for new inputs.

Reply:
I just pushed a change that does this and tests it. I believe all inputs are properly supported now, keeping in mind the restriction that epilogue vectorization is not yet handled correctly (#2169).